### ISSUE TITLE
How to ensure all top-level imports are sorted globally

### ISSUE THREAD

**Issue by user27182** [astral-sh/ruff#26783]
### Question

I'd like to keep all top-level imports sorted _and_ grouped together. I use `tool.ruff.lint.isort` for this, and `E402` to ensure all imports are at the top. But it seems lots of exceptions are baked into `E402` to allow for non-import code to be sprinkled in between actual imports. E.g. with isort and E402 enabled, this is apparently "correct", and ruff doesn't complain about the sorting:

``` py
import contextlib
from typing import get_args

with contextlib.suppress(ImportError):
    import numpy as np

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import numpy.typing as npt

from typing import get_origin
```

But the issue is that I want all `typing` imports grouped and sorted together in a single block, like this instead:

``` py
import contextlib
from typing import get_args
from typing import get_origin
from typing import TYPE_CHECKING

with contextlib.suppress(ImportError):
    import numpy as np

if TYPE_CHECKING:
    import numpy.typing as npt
```

Is there currently a way to enforce this?

For placing the `TYPE_CHECKING` block specifically at the bottom, there is https://github.com/astral-sh/ruff/issues/13976. But my issue also extends to other code like `with contextlib.suppress(ImportError):` messing up the sorting. Perhaps this should be placed below the imports as well? Or above is fine with me too. The main thing is to not break up the sorting, since this effectively creates two separate groups that are now sorted independently.

What happens in real code is that some modules have huge import lists, and maybe an import stops working (no longer exists in an old version of a dependency). So then the import may be suppressed by wrapping it with contextlib suppress:

E.g. this:
``` py
import b
import c
import d
```
becomes:
``` py
import b

with contexlib.suppress(ImportError):
    import c

import d
```

Except this silently and non-obviously breaks up the sorting into two separate groups. If someone later adds `import a`, it can easily be added to the "wrong" group, and we end up with the following "correctly sorted" imports:

``` py
import b

with contexlib.suppress(ImportError):
    import c

import a
import d
```

But really what I wanted was:

``` py
import a
import b
import d

with contexlib.suppress(ImportError):
    import c
```

Even then it's arguable this is not ideal, since users probably truly want _all_ imports, even suppressed ones, to be sorted together as a single contiguous group, such that `import c`, lands in the middle, _and_ `import a` lands at the top in all cases. But, if we assume the sorting is "dumb" and doesn't consider suppressed imports, then perhaps this is what I would settle for, with the `contextlib` code completely separate from the import code (i.e. move it to the top or the bottom such that all _other_ imports can be sorted together.) 

### Version

v0.15.15 (pre-commit)

**Comment by ntBre** [comment]
Interesting! Most of our recent reports about `E402` have actually been asking for more exceptions rather than removing exceptions. I think this might technically be a duplicate of #1251 because `isort --float-to-top` seems to produce the output you want:

```console
❯ isort --float-to-top - < try.py
import contextlib
from typing import TYPE_CHECKING, get_args, get_origin

with contextlib.suppress(ImportError):
    import numpy as np


if TYPE_CHECKING:
    import numpy.typing as npt
```

but changing something in `E402` could also be a possibility. 

In addition to https://github.com/astral-sh/ruff/issues/13976, I found https://github.com/astral-sh/ruff/issues/21132 and some other recent `E402` issues for reference: #20288 and #19904.

I'm starting to wonder if maybe a customizable list of exceptions would be the best way to go. You could set it to an empty list, and these other users could add their specific calls to it.

**Comment by user27182** [comment]
Thanks for investigating! 

> this might technically be a duplicate of https://github.com/astral-sh/ruff/issues/1251 because isort --float-to-top seems to produce the output you want

I didn't know about `isort --float-to-top`, but it seems this does indeed do exactly what I want. +1 for adding that feature! I'll leave it up to you if you think this issue should be closed accordingly.

> In addition to https://github.com/astral-sh/ruff/issues/13976, I found https://github.com/astral-sh/ruff/issues/21132 and some other recent E402 issues for reference: https://github.com/astral-sh/ruff/issues/20288 and https://github.com/astral-sh/ruff/issues/19904.

There is also a request for `warnings` [#21423](https://github.com/astral-sh/ruff/issues/21423) and `logging` [#19928](https://github.com/astral-sh/ruff/issues/19928) exceptions to E402

And your suggestion for a configurable exceptions list to `E402` seems reasonable. Even without making it configurable, maybe updating the docs to make it clear that `if TYPE_CHECKING` blocks and `with contextlib` branches are also exceptions to this rule? The docs currently only mention `os.environ` and `sys.path` exceptions:

> This rule makes an exception for both sys.path modifications (allowing for sys.path.insert, sys.path.append, etc.) and os.environ modifications between imports.

So it was a bit surprising to me that ruff didn't complain about `E402` here for the `get_origin` import not being at the top since there is an `if` check in between:

``` py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import numpy.typing as npt

from typing import get_origin
```

For this example it kind of makes sense, since there is at least an import inside the if block, but even this code doesn't cause ruff to complain:

``` py
from typing import TYPE_CHECKING

if True:
    x = 42

from typing import get_origin   # Not an E402 error, but maybe it should?
```
Seems like any flow control is probably just ignored completely? (`if`, `with`, `try...except`)? But rather than users guessing, maybe it's better to document this. And maybe point users to the not-yet-implemented `isort --float-to-top` feature to deal with this instead.
